### DISCUSSION TITLE
How to create  transformer to map repeated segments

### DISCUSSION THREAD

**Discussion by sal4455** [nextgenhealthcare/connect#5259] [Q&A]
Hi All,

I've created transformer for HL7V2 to FHIR mapping, I mapped almost every segments of ADT_A01 trigger event, but couldn't find a way to map repeated segments like AL1, OBX.
If there is any discussion , please share the link or help me out on how to map repeated segments in MirthConnect(3.11.0) .

 I'm using this code,let me know if it is correct or not :
<img width="277" alt="image" src="https://user-images.githubusercontent.com/104156028/176611304-96da394c-1312-4227-8129-9df7bcb088df.png">

on passing below hl7 message and using above mapping step, the output fhir json(Bundle resource) is creating 2 resources of AllergyIntolerance, but in both resources it is mapping values of first AL1 segment:
<img width="290" alt="image" src="https://user-images.githubusercontent.com/104156028/176612440-b947c49d-640e-40b1-8922-961f3fab5428.png">

**Comment by mjtrac** [comment]
When you are looking at transformers or filters, you will see a References tab at upper right.  You can search on that for "Iterate" or just type iter.  You should find some code templates that roughly show you this:

```
var myAL1_3_2s = Lists().list()

for each (seg in msg..AL1){
   myAL1_3_2s.append(seg['AL1.3']['AL1.3.2'].toString())
}

channelMap.put('myAL1_3_2s',myAL1_3_2s);
 
```


msg..AL1 with two dots generates a list of all segments named AL1 in the message msg.  The for each loop assigns each such segment as seg for one pass of the loop.  You start from seg and then add ['AL1.3'] to access the segment's third field, and ['AL1.3.2'] to access the second component of the third field.  You don't need to use a list, I'm just showing they can be captured in a list.  You could append to text, etc...  This can also be done for you via Iteration in the GUI -- see the User Guide.

**Reply by tonygermano** [reply]
To clarify, `msg.AL1` (one dot) and `msg..AL1` (two dots) will both generate a list of all segments named "AL1" in the message `msg`. One dot is the child operator, and two dots is the descendant operator. Because of the structure of HL7v2 xml, segment names are only ever found as direct children to the message root. Two dots takes slightly more time to produce the same result as instead of only checking segments, it needs to check every field, component, and subcomponent to see if they are named "AL1", which they won't be in an HL7 message.

**Comment by pacmano1** [comment]
Just a minor change from OP below,  concept seems fine.

```JS
var junk = Lists.list()
for each(al1 in msg..AL1)
{
    junk.add(
        {
            "setid": al1['AL1.1']['AL1.1.1'].toString(),
            "otherstuff": al1['AL1.4']['AL1.4.1'].toString()
        }
       )
}

$c('junk',junk)
```

**Comment by sal4455** [comment]
Thankyou @pacmano1 @mjtrac ,the above code worked.

Now, I'm mapping Resources id(AllergyIntolerance.id) to "fullurl" element of **Bundle Resource**.  I wrote below code to extract resource id :
<img width="324" alt="image" src="https://user-images.githubusercontent.com/104156028/176832512-5d3ce71a-9bbb-438f-bfd1-ac1ea33b9482.png">

and map FULLURL element of bundle resource using GUI:
<img width="354" alt="image" src="https://user-images.githubusercontent.com/104156028/176834125-c1c3ec5e-e245-435d-a4cd-b17b08152a06.png">

out of two AllergyIntolerance resources, 
In first resource, fullurl resource id is showing "undefined":
<img width="411" alt="image" src="https://user-images.githubusercontent.com/104156028/176834443-2c2292f4-db2c-4fb0-90d1-12f637ed1b51.png">

In second resource,fullurl element is not getting mapped:
<img width="367" alt="image" src="https://user-images.githubusercontent.com/104156028/176834785-b629c7d4-3415-49ea-b3b9-70d1888d862d.png">
